sdl2promela/stop_condition/translator.py

```diff
@@ -246,9 +299,185 @@ def _generate(expr: model.Selector):
     return result
 
 
-@dispatch(model.CallExpression)
-def _generate(expr: model.CallExpression):
-    raise TranslateException("Not implemented.")
+def _resolve_type(allTypes, t):
+    while t.kind == "ReferenceType":
+        if t.ReferencedTypeName not in allTypes:
+            raise ("Cannot resolve type {}".format(t.ReferencedTypeNam))
+        t = allTypes[t.ReferencedTypeName].type
+
+    return t
+
+
+@dispatch(GenerateContext, model.Selector)
+def _find_type(context: GenerateContext, selector: model.Selector):
+    index = 0
+    if isinstance(selector.elements[index], model.CallExpression):
+        finalType, allTypes = _find_type(context, selector.elements[index])
+        index = index + 1
+    else:
+        process_name = selector.elements[index].name
+        if process_name not in context.processes:
+            raise TranslateException(
+                "Cannot find process with name '{}'".format(process_name)
+            )
+
+        process = context.processes[process_name]
+        allTypes = getattr(process.DV, "types", {})
+
+        index = index + 1
+        variable_name = selector.elements[index].name
+
+        if variable_name not in process.variables:
+            raise TranslateException(
+                "Cannot find variable '{}' in process '{}'".format(
+                    variable_name, process_name
+                )
+            )
+        variable = process.variables[variable_name]
+
+        finalType = _resolve_type(allTypes, variable[0])
+        index = index + 1
+
+    while index < len(selector.elements):
+        member_name = selector.elements[index].name
+
+        if finalType.kind != "SequenceType":
+            raise TranslateException("Doom")
+
+        if member_name not in finalType.Children:
+            raise TranslateException(
+                "Unable to find member '{}' in '{}".format(member_name, finalType.CName)
+            )
+        member = finalType.Children[member_name]
+        finalType = _resolve_type(allTypes, member.type)
+        index = index + 1
+
+    return finalType, allTypes
+
+
+@dispatch(GenerateContext, model.CallExpression)
+def _find_type(context: GenerateContext, selector: model.CallExpression):
+    finalType, allTypes = _find_type(context, selector.function)
+
+    finalType = _resolve_type(allTypes, finalType)
+
+    if finalType.kind != "SequenceOfType":
+        raise TranslateException("Invalid access to SEQUENCE OF")
+
+    finalType = _resolve_type(allTypes, finalType.type)
```

Comment:
This looks like really_final_this_time_final_type... I would prefer a more descriptive name

Reply:
Changed